certbot/error_handler.py

```diff
@@ -34,29 +42,40 @@ class works best as a context manager. For example:
     if a signal is encountered, cleanup_func is called followed by the
     previously registered signal handler.
 
-    Every registered function is attempted to be run to completion
-    exactly once. If a registered function raises an exception, it is
-    logged and the next function is called. If a (different) handled
-    signal occurs while calling a registered function, it is attempted
-    to be called again by the next signal handler.
+    Each registered cleanup function is called exactly once. If a registered
+    function raises an exception, it is logged and the next function is called.
+    Signals received while the registered functions are executing are
+    deferred until they finish.
 
     """
     def __init__(self, func=None, *args, **kwargs):
+        self.body_executed = False
         self.funcs = []
         self.prev_handlers = {}
+        self.received_signals = []
         if func is not None:
             self.register(func, *args, **kwargs)
 
     def __enter__(self):
         self.set_signal_handlers()
 
     def __exit__(self, exec_type, exec_value, trace):
+        self.body_executed = True
+        retval = False
+        if exec_type is SignalExit:
+            logger.debug("Encountered signals: %s", self.received_signals)
+            self.call_registered()
+            for signum in self.received_signals:
+                self.call_signal(signum)
```

Comment:
I think we should call `_reset_signal_handlers()` before doing this. There are two benefits to this approach:

1. We're not looping over a list that may be modified if we get another signal.
2. We ensure no signals are incorrectly ignored (what currently happens if we get a signal after this loop but before the call to `_reset_signal_handlers`?).